ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java

```diff
@@ -123,14 +124,14 @@ public boolean hasNext() {
     @Override
     public void remove() {
       replicaIterator.remove();
-      if (inflightCount >= parallelism) {
+      if (inflightCount >= parallelism && unexpiredRequestSendTimes.size() > 0) {
         // we are here because oldest request is past due
-        Map.Entry<ReplicaId, Long> oldestEntry = unexpiredRequestSendTimes.entrySet().iterator().next();
-        expiredRequestSendTimes.put(oldestEntry.getKey(), oldestEntry.getValue());
+        Map.Entry<ReplicaId, Pair<Boolean, Long>> oldestEntry = unexpiredRequestSendTimes.entrySet().iterator().next();
+        expiredRequestSendTimes.put(oldestEntry.getKey(), oldestEntry.getValue().getSecond());
         unexpiredRequestSendTimes.remove(oldestEntry.getKey());
         pastDueCounter.inc();
       }
-      unexpiredRequestSendTimes.put(lastReturned, time.milliseconds());
+      unexpiredRequestSendTimes.put(lastReturned, new Pair<>(false, time.milliseconds()));
```

Comment:
The semantics for this iterator seems to be that "remove()" indicates that the `ReplicaId` will be used to send requests to - in order to respect that semantic, it has to be this way.